Nucleation options

fvitt opened the pull request:
Mike has introduced an alternative nucleation method option

fvitt committed:
initial version from Mike

        modified:   sulfnucrate.F90

fvitt committed:
add option to select nucleation method

        modified:   carma_globaer.h
        modified:   carma_mod.F90
        modified:   carma_types_mod.F90
        modified:   sulfnucrate.F90

fvitt committed:
code clean up

        modified:   sulfnucrate.F90

fvitt commented:
Standalone CARMA regression tests all pass

tilmes commented:
@fvitt does this include a namelist variable for the nucleation scheme yet?

fvitt commented:
@tilmes I have changes to CAM for the namelist option for the nucleation scheme.

tilmes commented:
@fvitt OK

fvitt commented on sulfnucrate.F90:
```diff
@@ -4,51 +4,386 @@
+  real(kind=f)      :: temp_bb          ! bounded value of temperature in Kelvins
+  real(kind=f)      :: rh_bb            ! bounded value of relative humidity
+
+  integer, parameter :: newnuc_method_flagaa = 1
+  !  1=Zhao & Turco (1995) binary
+  !  2=vehkamaki et al (2002) binary
+
```
remove these lines

cbardeen replied:
These lines shouldn't been in here. Also in CARMA all the enumerations are in carma_enums_mod.F90, and typically have been numbers not strings. I don't really care if they are numbers or strings, but they should be consistent to make it easier to read the code and find things.

cbardeen commented on carma_globaer.h:
```diff
@@ -187,14 +187,15 @@
 #define gsticki       carma%f_gsticki
 #define gstickl       carma%f_gstickl
 #define cstick        carma%f_cstick
+#define nucl_method   carma%nucl_method
```
nucl_method may not be a good term to use as this is only being used for sulfate nucleation and there is already a field (inucproc) that is used to determine the type of nucleation that is used for clouds and aerosols and has values in carma_enums_mod.F90.


 !!  Define values of flag used for nucleation process
  !!  specification (inucproc).
  !!
  !!  NOTE: Some of these can be used in combination, so for aerosol freezing this is treated
  !!  as a bit mask. When setting for one (or more) of the Aerosol freezing methods, use:
  !!    IAERFREEZE + I_AF_xxx + I_AF_yyy + ...
  integer, public, parameter :: I_AF_TABAZADEH_2000 = 1     !! Aerosol Freezing, Tabazadeh[2000]
  integer, public, parameter :: I_AF_KOOP_2000      = 2     !! Aerosol Freezing, Koop[2000]
  integer, public, parameter :: I_AF_MOHLER_2010    = 4     !! Aerosol Freezing, Mohler[2010]
  integer, public, parameter :: I_AF_MURRAY_2010    = 8     !! Glassy Aerosol Freezing, Murray[2010]
  integer, public, parameter :: I_DROPACT           = 256   !! Droplet Activation
  integer, public, parameter :: I_AERFREEZE         = 512   !! Aerosol Freezing
  integer, public, parameter :: I_DROPFREEZE        = 1024  !! Droplet Freezing
  integer, public, parameter :: I_ICEMELT           = 2048  !! Ice Melting
  integer, public, parameter :: I_HETNUC            = 4096  !! Heterogeneous Nucleation
  integer, public, parameter :: I_HOMNUC            = 8192  !! Binary homogeneous gas-to-particle nucleation
  integer, public, parameter :: I_HETNUCSULF        = 16384 !! Binary homogeneous gas-to-particle nucleation

cbardeen commented on sulfnucrate.F90:
```diff
@@ -4,51 +4,386 @@
 !!  Calculates particle production rates due to nucleation <rhompe>:
 !!  binary homogeneous nucleation of sulfuric acid and water only
-!!  Numerical method follows Zhao & Turco, JAS, V.26, No.5, 1995.
+!!  Numerical method follows one of the following:
```
I think it would be better to break this up into 3 files following the pattern of vaporp, where there is a top level routine (vaporp.F90) that has the switch statement to select the appropriate routine, and then each routine is in its own file. So I would suggest sulfnucrate_zhao1995.F90 and sulfnucrate_vehkamaki2002.F90 as the new files. This keeps the files simpler as we are likely to add more of these in the future.

You might change the subroutine names to match the file names as this is typical in the base code and makes it easier when searching the code.

cbardeen commented on sulfnucrate.F90:
```diff
@@ -4,51 +4,386 @@
-  integer, intent(out)                 :: nucbin      !! bin in which nucleation occurs
-  real(kind=f), intent(out)            :: nucrate     !! nucleation rate #/x/y/z/s
-  integer, intent(inout)               :: rc          !! return code, negative indicates failure
+  real(kind=f), intent(out)            :: radius_cluster !! critical radius (cm)
```
Why was name change from rstar to radius cluster? It is rstar in the place that calls this - sulfnuc.F90. I am not a fan of changing the parameter names in the middle as it makes searches harder. I am okay if zhao or vehkamaki want to have different terminology in their routines, but would rather have it be te same before that.

cbardeen commented on sulfnucrate.F90:
```diff
@@ -4,51 +4,386 @@
 
 !!  Calculates particle production rates due to nucleation <rhompe>:
```
in sulfnuc.F90, it does homogeneous nucleation first and then uses the rstar value with the heterogeneous nucleation. I don't know if this is still fine when zhao is replaced by vehkamaki or not or whether the rstar needs to be calculated as part of the heterogeneous nucleation path. This is a question for Mike. However, in the stand-alone there needs to be two tests added to exercise this code. They can be copies of the carma_sulfhettest and the carma_sulfatetest, which the nucleation method switched from zhao to vehkamaki.

cbardeen commented on sulfnucrate.F90:
```diff
@@ -4,51 +4,386 @@
+  nucrate_cgs = 0.0_f
+  radius_cluster = 0.0_f
+  mass_cluster_dry = 0.0_f
+  ftry = -99999._f
```
ftry does not get calculated when using Vehkamaki, but is used in heterogeneous nucleation. If the value is -99999, then I think it will shut off all heterogeneous nucleation.

  ftry1 = ftry * fv1                                              
!  ftry1 = ftry * fh                                              
  if (ftry1 .lt. -1000._f) then
    nucrate = 0._f
  else


However, it doesn't look like the trop_strat model uses heterogeneous nucleation. So it may not be an issue running that model. Is the reason we don't do heterogeneous nucleation because we don't have a dust group?

fvitt committed:
merge to head of ESCOMP/main

fvitt committed:
Mikes latest changes -- check ftry

       modified:   carma_constants_mod.F90
       modified:   sulfhetnucrate.F90
       modified:   sulfnucrate.F90

fvitt committed:
set default nucl_method = 'NONE'

       modified:   carma_types_mod.F90

fvitt committed:
rename "nucl_method" as "sulf_nucl_method"

       modified:   carma_globaer.h
       modified:   carma_mod.F90
       modified:   carma_types_mod.F90
       modified:   sulfnucrate.F90

fvitt committed:
Changes needed for CAM

        modified:   carma_constants_mod.F90
        modified:   carma_globaer.h
        modified:   carma_mod.F90
        modified:   carma_types_mod.F90
        modified:   sulfnucrate.F90

tilmes reviewed (commented):
What does it mean if nucl_method is NONE? Doesn't CARMA need a nucleation method to work?

fvitt commented:
> What does it mean if nucl_method is NONE? Doesn't CARMA need a nucleation method to work?

This just means the nucl_method needs to be explicitly set for sulfate nucleation to work.  For CAM the default namelist setting is 'ZhaoTurco'.

cbardeen commented on sulfhetnucrate.F90:
```diff
@@ -17,24 +17,24 @@ subroutine sulfhetnucrate(carma, cstate, iz, igroup, nucbin, h2o, h2so4, beta1,
   real(kind=f), intent(in)             :: beta2
   real(kind=f), intent(in)             :: ftry
-  real(kind=f), intent(in)             :: rstar       !! critical radius (cm)
+  real(kind=f), intent(inout)          :: rstar       !! critical radius (cm)
```
Why is this inout? It isn't needed outside of this routine is it? Would you just need to make a local variable for rstar so you could calculate it when not supplied?

I was hoping Mike would pull out the part of Zhao that claculates ftry and rstar rather than putting a call to the whole routine here. THen you could probably do it all the time and not have ftry and rstar passed in here. Still seems kind of fragile if we have more routines coming in the future.

cbardeen commented on carma_types_mod.F90:
```diff
@@ -509,6 +509,18 @@ module carma_types_mod
     real(kind=f), allocatable, dimension(:)          :: f_dwave      ! (NWAVE)
     logical, allocatable, dimension(:)               :: f_do_wave_emit  ! (NWAVE)
+
+    !  Sulfate Nucleation Methods:
```
How is this being used? Does Vehkamaki have heterogeneous nucleation? What would you set this for if you wanted Zhao for heterogenous and Vehkamki for homogeneous? Is this really a specifier for the homogeneous nucleation method and not the heterogenous nucleation method? It might be good to rename this variable or at least add comments, if it only controls the homogeneous nucleation.

fvitt committed:
Mike's clean up changes

       modified:   sulfhetnucrate.F90
       modified:   sulfnuc.F90
       modified:   sulfnucrate.F90

fvitt committed:
update ChangeLog
